source/material_model/rheology/compositional_viscosity_prefactors.cc

```diff
@@ -0,0 +1,170 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/rheology/compositional_viscosity_prefactors.h>
+#include <aspect/utilities.h>
+#include <aspect/global.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/parameter_handler.h>
+#include <aspect/simulator_signals.h>
+#include <aspect/parameters.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace Rheology
+    {
+      template <int dim>
+      CompositionalViscosityPrefactors<dim>::CompositionalViscosityPrefactors ()
+        = default;
+
+      template <int dim>
+      std::vector<double>
+      CompositionalViscosityPrefactors<dim>::compute_viscosities (const MaterialModel::MaterialModelInputs<dim> &in,
+                                                                  const double base_viscosity,
+                                                                  const unsigned int composition_index,
+                                                                  const unsigned int q) const
+      {
+        std::vector<double> factored_viscosities(number_of_prefactors, base_viscosity);
+        switch (viscosity_prefactor_scheme)
+          {
+            case none:
+            {
+              break;
+            }
+            case hk04_olivine_hydration:
+            {
+              // We calculate the atomic H/Si ppm (C_OH) at each point to compute the water fugacity of
+              // olivine assuming a composition of 90 wt% Forsterite and 10 wt% Fayalite from Hirth
+              // and Kohlstaedt 2004 10.1029/138GM06.
+              const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+              const double weight_fraction_H2O = in.composition[q][bound_fluid_idx]; // mass fraction of bound water
+              const double weight_fraction_olivine = 1 - weight_fraction_H2O; // mass fraction of olivine
+              const double COH = (weight_fraction_H2O/molar_mass_H2O) / (weight_fraction_olivine/molar_mass_olivine) * 1e6; // COH in H / Si ppm
+              const double point_water_fugacity = COH / A_H2O * \
+                                                  std::exp((activation_energy_H2O + in.pressure[q]*activation_volume_H2O)/ \
+                                                           (constants::gas_constant * in.temperature[q]));
+
+              factored_viscosities[0] = base_viscosity*std::pow(point_water_fugacity, -diffusion_water_fugacity_exponents[composition_index]);
+              factored_viscosities[1] = base_viscosity*std::pow(point_water_fugacity, -dislocation_water_fugacity_exponents[composition_index]);
```

Comment:
I have another suggestion in addition to what @bobmyhill and @naliboff have already said. 
This code implicitly says that diffusion creep has index zero, and dislocation creep has index 1. That's not really intuitive when only looking at it from the material/rheology model that uses your new rheology. I suggest something like this:

Make an enum in your rheology, something like

```
struct FlowLaw
    {
      enum Kind
      {
        diffusion,
        dislocation
      };
    };
```

Then make a little function in your rheology `get_flow_law_index(typename FlowLaw::Kind flow_law)` that returns the index as an unsigned int (0 for diffusion, 1 for dislocation). 

Then in visco_plastic.cc you can do something like

```
diffusion_index = compositional_viscosity_prefactors.get_flow_law_index(FlowLaw::diffusion);
non_yielding_viscosity = compositional_viscosity_prefactors.compute_viscosities(in, viscosity_diffusion, j, i)[diffusion_index];
```

@bobmyhill Does that seem reasonable?